{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/paulynamagana/AFDB_notebooks/blob/main/AFDB_FTP.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "<img src = \"https://www.embl.org/about/info/communications/wp-content/uploads/2017/09/Ebi_official_logo.png\"\n",
        " height=\"100\" align=\"right\">\n",
        "\n",
        "# Access structures from AlphaFold DB via FTP\n",
        "\n",
        "FTP, or File Transfer Protocol, is a standard network protocol facilitating the exchange of files between computers.\n",
        "\n",
        "\n",
        "\n",
        "<br>\n",
        "\n",
        "As of September 2023, the EMBL-EBIâ€™s FTP area hosts TAR files for proteomes of 48 organisms, including model organisms and WHO pathogens of interest.\n",
        "\n",
        "We document every data version update in our [CHANGELOG](https://ftp.ebi.ac.uk/pub/databases/alphafold/CHANGELOG.txt).\n",
        "\n",
        "\n",
        "<br>\n",
        "\n",
        "The folders are named following a structured convention, comprising three distinct elements separated by underscores:\n",
        "\n",
        "- Reference Proteome (UPID): UP000000429\n",
        "- Taxonomy ID: 85962\n",
        "- Organism: HELPY (derived from the first three characters of the genus, \"Helicobacter,\" and the first two characters of the species, \"pylori\").\n",
        "\n",
        "<br>\n",
        "\n",
        "In order to understand the folders, visit the [Downloads tab](https://alphafold.ebi.ac.uk/download)\n",
        "\n",
        "You can also find the compressed files for Swiss-Prot which contains 542,378 predicted structures:\n",
        "\n",
        "|File type|File name|Size|\n",
        "|---------|--------------|---------------------|\n",
        "|Swiss-Prot (CIF Files)|swissprot_cif_v4.tar| 37,643 MB|\n",
        "|Swiss-Prot (PDB files)|swissprot_pdb_v4.tar|26,935 MB|\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "LDa23wtBDv5w"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bZnvx4lyvZIj",
        "outputId": "782515d4-3878-4c20-94b8-5a56b42c5277",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CHANGELOG.txt\n",
            "Outreach\n",
            "README.txt\n",
            "Training\n",
            "accession_ids.csv\n",
            "diffs.ndjson.gz\n",
            "download_metadata.json\n",
            "latest\n",
            "proteomes\n",
            "sequences.fasta\n",
            "v1\n",
            "v2\n",
            "v3\n",
            "v4\n",
            "v5\n",
            "v6\n"
          ]
        }
      ],
      "source": [
        "#@title #Run to see what's in the FTP area\n",
        "#@markdown Run this block to see what's in the FTP area\n",
        "import ftplib\n",
        "import io\n",
        "import json\n",
        "from ftplib import FTP\n",
        "import tarfile\n",
        "import tempfile\n",
        "import os\n",
        "from google.colab import files\n",
        "import zipfile\n",
        "\n",
        "ftp_server = ftplib.FTP(\"ftp.ebi.ac.uk\")\n",
        "\n",
        "# Login as an anonymous user\n",
        "ftp_server.login(\"anonymous\", \"anonymous@\")\n",
        "\n",
        "# Navigate to the directory\n",
        "ftp_server.cwd(\"/pub/databases/alphafold/\")\n",
        "\n",
        "# List the contents of the directory\n",
        "# Retrieve and print the list of file names in the directory\n",
        "file_names = ftp_server.nlst()\n",
        "for file_name in file_names:\n",
        "    print(file_name)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #See all the available proteomes in one version\n",
        "#@markdown This block will retrieve a list of the files inside the version archive you define\n",
        "\n",
        "folder_navigate = \"v6\" #@param {type:\"string\"}\n",
        "#@markdown `folder_navigate` is the version of the AFDB that you want to use to download\n",
        "\n",
        "#@title Navigate to the \"v4\" directory\n",
        "ftp_server.cwd(folder_navigate)\n",
        "\n",
        "# Retrieve and print the list of file names in the directory\n",
        "file_names = ftp_server.nlst()\n",
        "for file_name in file_names:\n",
        "    print(file_name)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OSaz9_sLwrhN",
        "outputId": "6331cc98-5937-4934-abdd-7934c3683fc8",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "UP000000429_85962_HELPY_v6.tar\n",
            "UP000000437_7955_DANRE_v6.tar\n",
            "UP000000535_242231_NEIG1_v6.tar\n",
            "UP000000559_237561_CANAL_v6.tar\n",
            "UP000000579_71421_HAEIN_v6.tar\n",
            "UP000000586_171101_STRR6_v6.tar\n",
            "UP000000589_10090_MOUSE_v6.tar\n",
            "UP000000625_83333_ECOLI_v6.tar\n",
            "UP000000799_192222_CAMJE_v6.tar\n",
            "UP000000803_7227_DROME_v6.tar\n",
            "UP000000805_243232_METJA_v6.tar\n",
            "UP000000806_272631_MYCLE_v6.tar\n",
            "UP000001014_99287_SALTY_v6.tar\n",
            "UP000001450_36329_PLAF7_v6.tar\n",
            "UP000001584_83332_MYCTU_v6.tar\n",
            "UP000001631_447093_AJECG_v6.tar\n",
            "UP000001940_6239_CAEEL_v6.tar\n",
            "UP000002059_502779_PARBA_v6.tar\n",
            "UP000002195_44689_DICDI_v6.tar\n",
            "UP000002296_353153_TRYCC_v6.tar\n",
            "UP000002311_559292_YEAST_v6.tar\n",
            "UP000002438_208964_PSEAE_v6.tar\n",
            "UP000002485_284812_SCHPO_v6.tar\n",
            "UP000002494_10116_RAT_v6.tar\n",
            "UP000002716_300267_SHIDS_v6.tar\n",
            "UP000005640_9606_HUMAN_v6.tar\n",
            "UP000006304_1133849_9NOCA1_v6.tar\n",
            "UP000006548_3702_ARATH_v6.tar\n",
            "UP000006672_6279_BRUMA_v6.tar\n",
            "UP000007305_4577_MAIZE_v6.tar\n",
            "UP000007841_1125630_KLEPH_v6.tar\n",
            "UP000008153_5671_LEIIN_v6.tar\n",
            "UP000008524_185431_TRYB2_v6.tar\n",
            "UP000008816_93061_STAA8_v6.tar\n",
            "UP000008827_3847_SOYBN_v6.tar\n",
            "UP000008854_6183_SCHMA_v6.tar\n",
            "UP000018087_1391915_SPOS1_v6.tar\n",
            "UP000024404_6282_ONCVO_v6.tar\n",
            "UP000030665_36087_TRITR_v6.tar\n",
            "UP000035681_6248_STRER_v6.tar\n",
            "UP000053029_1442368_9EURO2_v6.tar\n",
            "UP000059680_39947_ORYSJ_v6.tar\n",
            "UP000078237_100816_9PEZI1_v6.tar\n",
            "UP000094526_86049_9EURO1_v6.tar\n",
            "UP000270924_6293_WUCBA_v6.tar\n",
            "UP000274756_318479_DRAME_v6.tar\n",
            "swissprot_cif_v6.tar\n",
            "swissprot_pdb_v6.tar\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #Get the metadata\n",
        "#@markdown This block will retrieve the metadata and print it, you can see the tar file, the specie, common name and metadata accompanying this.\n",
        "\n",
        "# FTP server details\n",
        "ftp_server = \"ftp.ebi.ac.uk\"\n",
        "\n",
        "try:\n",
        "    with ftplib.FTP(ftp_server) as ftp:\n",
        "        print(\"Accessing metadata...\")\n",
        "        ftp.login(user=\"anonymous\", passwd=\"anonymous\")\n",
        "        ftp.cwd(\"/pub/databases/alphafold\")  # Navigate to the directory containing the metadata file\n",
        "\n",
        "        with io.BytesIO() as bio:\n",
        "            # Attempt to download the metadata file\n",
        "            ftp.retrbinary('RETR download_metadata.json', bio.write)\n",
        "            bio.seek(0)  # Go to the start of the BytesIO buffer\n",
        "            metadata = json.load(bio)\n",
        "\n",
        "    # Assuming metadata is a list of dictionaries, similar to the sample above\n",
        "    if metadata:\n",
        "        # Print the headers\n",
        "        headers = metadata[0].keys()\n",
        "        print(\"\\t\".join(headers))\n",
        "\n",
        "        # Print the values for each record\n",
        "        for record in metadata:\n",
        "            values = [str(record[key]) for key in headers]\n",
        "            print(\"\\t\".join(values))\n",
        "    else:\n",
        "        print(\"No metadata found.\")\n",
        "\n",
        "except Exception as e:\n",
        "    print(f\"Failed to fetch metadata: {e}\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "hie6E6PUx9di",
        "outputId": "f7d1d4f5-526d-4598-f424-d51113d6bd08",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Accessing metadata...\n",
            "archive_name\tspecies\tcommon_name\tlatin_common_name\treference_proteome\tnum_predicted_structures\tsize_bytes\ttype\n",
            "UP000006548_3702_ARATH_v6.tar\tArabidopsis thaliana\tArabidopsis\tTrue\tUP000006548\t27402\t3877400576\tproteome\n",
            "UP000001940_6239_CAEEL_v6.tar\tCaenorhabditis elegans\tNematode worm\tFalse\tUP000001940\t19700\t2777489920\tproteome\n",
            "UP000000559_237561_CANAL_v6.tar\tCandida albicans\tC. albicans\tTrue\tUP000000559\t5973\t1028541440\tproteome\n",
            "UP000000437_7955_DANRE_v6.tar\tDanio rerio\tZebrafish\tFalse\tUP000000437\t26290\t4979466752\tproteome\n",
            "UP000002195_44689_DICDI_v6.tar\tDictyostelium discoideum\tDictyostelium\tTrue\tUP000002195\t12612\t2292386816\tproteome\n",
            "UP000000803_7227_DROME_v6.tar\tDrosophila melanogaster\tFruit fly\tFalse\tUP000000803\t13461\t2319755776\tproteome\n",
            "UP000000625_83333_ECOLI_v6.tar\tEscherichia coli\tE. coli\tTrue\tUP000000625\t4370\t477871104\tproteome\n",
            "UP000008827_3847_SOYBN_v6.tar\tGlycine max\tSoybean\tFalse\tUP000008827\t55796\t7616606720\tproteome\n",
            "UP000005640_9606_HUMAN_v6.tar\tHomo sapiens\tHuman\tFalse\tUP000005640\t23586\t5177506304\tproteome\n",
            "UP000000805_243232_METJA_v6.tar\tMethanocaldococcus jannaschii\tM. jannaschii\tTrue\tUP000000805\t1773\t181713920\tproteome\n",
            "UP000000589_10090_MOUSE_v6.tar\tMus musculus\tMouse\tFalse\tUP000000589\t21452\t3782078976\tproteome\n",
            "UP000059680_39947_ORYSJ_v6.tar\tOryza sativa\tAsian rice\tFalse\tUP000059680\t43645\t4723294720\tproteome\n",
            "UP000002494_10116_RAT_v6.tar\tRattus norvegicus\tRat\tFalse\tUP000002494\t22152\t3776509440\tproteome\n",
            "UP000002311_559292_YEAST_v6.tar\tSaccharomyces cerevisiae\tBudding yeast\tFalse\tUP000002311\t6055\t1024019456\tproteome\n",
            "UP000002485_284812_SCHPO_v6.tar\tSchizosaccharomyces pombe\tFission yeast\tFalse\tUP000002485\t5196\t840982016\tproteome\n",
            "UP000007305_4577_MAIZE_v6.tar\tZea mays\tMaize\tFalse\tUP000007305\t39139\t5023941632\tproteome\n",
            "UP000001631_447093_AJECG_v6.tar\tAjellomyces capsulatus\tAjellomyces capsulatus\tFalse\tUP000001631\t9199\t1428343296\tglobal_health\n",
            "UP000006672_6279_BRUMA_v6.tar\tBrugia malayi\tBrugia malayi\tFalse\tUP000006672\t10972\t1713548288\tglobal_health\n",
            "UP000000799_192222_CAMJE_v6.tar\tCampylobacter jejuni\tC. jejuni\tTrue\tUP000000799\t1620\t182684672\tglobal_health\n",
            "UP000094526_86049_9EURO1_v6.tar\tCladophialophora carrionii\tCladophialophora carrionii\tFalse\tUP000094526\t11170\t1812907520\tglobal_health\n",
            "UP000274756_318479_DRAME_v6.tar\tDracunculus medinensis\tDracunculus medinensis\tFalse\tUP000274756\t10834\t1429495808\tglobal_health\n",
            "UP000053029_1442368_9EURO2_v6.tar\tFonsecaea pedrosoi\tFonsecaea pedrosoi\tFalse\tUP000053029\t12509\t2111254016\tglobal_health\n",
            "UP000000579_71421_HAEIN_v6.tar\tHaemophilus influenzae\tH. influenzae\tTrue\tUP000000579\t1660\t182781952\tglobal_health\n",
            "UP000000429_85962_HELPY_v6.tar\tHelicobacter pylori\tH. pylori\tTrue\tUP000000429\t1540\t173476864\tglobal_health\n",
            "UP000007841_1125630_KLEPH_v6.tar\tKlebsiella pneumoniae\tK. pneumoniae\tTrue\tUP000007841\t5727\t585231360\tglobal_health\n",
            "UP000008153_5671_LEIIN_v6.tar\tLeishmania infantum\tL. infantum\tTrue\tUP000008153\t7924\t1580325376\tglobal_health\n",
            "UP000078237_100816_9PEZI1_v6.tar\tMadurella mycetomatis\tMadurella mycetomatis\tFalse\tUP000078237\t9561\t1610682368\tglobal_health\n",
            "UP000000806_272631_MYCLE_v6.tar\tMycobacterium leprae\tMycobacterium leprae\tFalse\tUP000000806\t1602\t184752128\tglobal_health\n",
            "UP000001584_83332_MYCTU_v6.tar\tMycobacterium tuberculosis\tM. tuberculosis\tTrue\tUP000001584\t3991\t449242624\tglobal_health\n",
            "UP000000535_242231_NEIG1_v6.tar\tNeisseria gonorrhoeae\tN. gonorrhoeae\tTrue\tUP000000535\t2106\t204283904\tglobal_health\n",
            "UP000006304_1133849_9NOCA1_v6.tar\tNocardia brasiliensis\tNocardia brasiliensis\tFalse\tUP000006304\t8398\t914921472\tglobal_health\n",
            "UP000024404_6282_ONCVO_v6.tar\tOnchocerca volvulus\tOnchocerca volvulus\tFalse\tUP000024404\t12039\t1698997248\tglobal_health\n",
            "UP000002059_502779_PARBA_v6.tar\tParacoccidioides lutzii\tParacoccidioides lutzii\tFalse\tUP000002059\t8794\t1356795904\tglobal_health\n",
            "UP000001450_36329_PLAF7_v6.tar\tPlasmodium falciparum\tP. falciparum\tTrue\tUP000001450\t5168\t1203016192\tglobal_health\n",
            "UP000002438_208964_PSEAE_v6.tar\tPseudomonas aeruginosa\tP. aeruginosa\tTrue\tUP000002438\t5555\t641853440\tglobal_health\n",
            "UP000001014_99287_SALTY_v6.tar\tSalmonella typhimurium\tS. typhimurium\tTrue\tUP000001014\t4526\t499788800\tglobal_health\n",
            "UP000008854_6183_SCHMA_v6.tar\tSchistosoma mansoni\tSchistosoma mansoni\tFalse\tUP000008854\t9735\t1889169408\tglobal_health\n",
            "UP000002716_300267_SHIDS_v6.tar\tShigella dysenteriae\tS. dysenteriae\tTrue\tUP000002716\t3893\t390678016\tglobal_health\n",
            "UP000018087_1391915_SPOS1_v6.tar\tSporothrix schenckii\tSporothrix schenckii\tFalse\tUP000018087\t8652\t1592406528\tglobal_health\n",
            "UP000008816_93061_STAA8_v6.tar\tStaphylococcus aureus\tS. aureus\tTrue\tUP000008816\t2888\t286346752\tglobal_health\n",
            "UP000000586_171101_STRR6_v6.tar\tStreptococcus pneumoniae\tS. pneumoniae\tTrue\tUP000000586\t2031\t211535872\tglobal_health\n",
            "UP000035681_6248_STRER_v6.tar\tStrongyloides stercoralis\tStrongyloides stercoralis\tFalse\tUP000035681\t15335\t2928591360\tglobal_health\n",
            "UP000030665_36087_TRITR_v6.tar\tTrichuris trichiura\tTrichuris trichiura\tFalse\tUP000030665\t9563\t1427457536\tglobal_health\n",
            "UP000008524_185431_TRYB2_v6.tar\tTrypanosoma brucei\tTrypanosoma brucei\tFalse\tUP000008524\t8491\t1409980416\tglobal_health\n",
            "UP000002296_353153_TRYCC_v6.tar\tTrypanosoma cruzi\tT. cruzi\tTrue\tUP000002296\t19036\t3102152192\tglobal_health\n",
            "UP000270924_6293_WUCBA_v6.tar\tWuchereria bancrofti\tWuchereria bancrofti\tFalse\tUP000270924\t12725\t1486524928\tglobal_health\n",
            "Failed to fetch metadata: 'species'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #Extract specific file type (mmCIF or PDB) for all fragments for a UniProt accession\n",
        "#@markdown This block will download all the fragments for a specific UniProt accession <br>\n",
        "#@markdown <strong>NOTE:</strong>  You will get a pop-up window to download files to your local computer\n",
        "\n",
        "\n",
        "#Input parameters\n",
        "database_version = 'v6' #@param {type:\"string\"}\n",
        "tar_file = 'UP000005640_9606_HUMAN_v6.tar' #@param {type:\"string\"}\n",
        "#@markdown Make sure this file coincides with the database version you're searching in (i.e. end of tar file) It should say `v4`  if you're searching in the `version 4` within the database\n",
        "UniProt_accession = 'Q8WXH0'  #@param {type:\"string\"}\n",
        "file_type = \"pdb\" #@param {type:\"string\"}\n",
        "#@markdown If you wish to download `cif` or `pdb` files, make sure it doesn't contain whitespaces\n",
        "\n",
        "def extract_files_and_zip_from_ftp(folder_navigate, tar_file, file_fragment_name, file_type):\n",
        "    username = 'anonymous'\n",
        "    password = 'anonymous'\n",
        "    ftp_server = \"ftp.ebi.ac.uk\"\n",
        "    base_path = f'pub/databases/alphafold/{folder_navigate}/'\n",
        "    file_path = base_path + tar_file\n",
        "    extracted_files = []  # To keep track of extracted file names\n",
        "\n",
        "    # Connect to the FTP server\n",
        "    with FTP(ftp_server) as ftp:\n",
        "        ftp.login(username, password)\n",
        "\n",
        "        # Use a temporary file to store the tar file\n",
        "        with tempfile.NamedTemporaryFile(delete=False) as tmp_file:\n",
        "            try:\n",
        "                ftp.retrbinary(f'RETR {file_path}', tmp_file.write)\n",
        "                tar_file_path = tmp_file.name\n",
        "            except Exception as e:\n",
        "                print(f\"Error downloading the file: {e}\")\n",
        "                return\n",
        "\n",
        "        # Open the temporary tar file for reading\n",
        "        try:\n",
        "            with tarfile.open(tar_file_path, mode=\"r:*\") as tar:\n",
        "                # Search through the entire tar file for all matches and check filename contains the file fragment and has a .cif extension\n",
        "                for member in tar.getmembers():\n",
        "                    if file_fragment_name in member.name and member.name.endswith(f'.{file_type.lower()}.gz'):\n",
        "                        extracted_file = tar.extractfile(member)\n",
        "                        if extracted_file:\n",
        "                            content = extracted_file.read()\n",
        "                            output_filename = f\"{member.name.replace('/', '_')}\"\n",
        "                            with open(output_filename, 'wb') as f_out:\n",
        "                                f_out.write(content)\n",
        "                            print(f\"Extracted and saved {member.name} as {output_filename}.\")\n",
        "                            extracted_files.append(output_filename)\n",
        "        except tarfile.TarError as e:\n",
        "            print(f\"Error reading the tar file: {e}\")\n",
        "        finally:\n",
        "            # Clean up the temporary file\n",
        "            os.remove(tar_file_path)\n",
        "\n",
        "    # Zip the extracted files\n",
        "    zip_filename = \"extracted_files.zip\"\n",
        "    with zipfile.ZipFile(zip_filename, 'w') as zipf:\n",
        "        for file in extracted_files:\n",
        "            zipf.write(file)\n",
        "            os.remove(file)  # Optional: remove the file after adding it to the zip to save space\n",
        "    print(f\"Created zip archive: {zip_filename}\")\n",
        "\n",
        "    # Download the zip file\n",
        "    files.download(zip_filename)\n",
        "\n",
        "\n",
        "\n",
        "extract_files_and_zip_from_ftp(database_version, tar_file, UniProt_accession, file_type)\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "2WQ9BEmwP7RK",
        "outputId": "c6e5bc11-3cac-4b36-bc7e-5ff44a7d1396",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 538
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extracted and saved AF-Q8WXH0-F1-model_v6.pdb.gz as AF-Q8WXH0-F1-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F10-model_v6.pdb.gz as AF-Q8WXH0-F10-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F11-model_v6.pdb.gz as AF-Q8WXH0-F11-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F12-model_v6.pdb.gz as AF-Q8WXH0-F12-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F13-model_v6.pdb.gz as AF-Q8WXH0-F13-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F14-model_v6.pdb.gz as AF-Q8WXH0-F14-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F15-model_v6.pdb.gz as AF-Q8WXH0-F15-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F16-model_v6.pdb.gz as AF-Q8WXH0-F16-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F17-model_v6.pdb.gz as AF-Q8WXH0-F17-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F18-model_v6.pdb.gz as AF-Q8WXH0-F18-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F19-model_v6.pdb.gz as AF-Q8WXH0-F19-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F2-model_v6.pdb.gz as AF-Q8WXH0-F2-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F20-model_v6.pdb.gz as AF-Q8WXH0-F20-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F21-model_v6.pdb.gz as AF-Q8WXH0-F21-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F22-model_v6.pdb.gz as AF-Q8WXH0-F22-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F23-model_v6.pdb.gz as AF-Q8WXH0-F23-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F24-model_v6.pdb.gz as AF-Q8WXH0-F24-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F25-model_v6.pdb.gz as AF-Q8WXH0-F25-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F26-model_v6.pdb.gz as AF-Q8WXH0-F26-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F27-model_v6.pdb.gz as AF-Q8WXH0-F27-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F28-model_v6.pdb.gz as AF-Q8WXH0-F28-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F29-model_v6.pdb.gz as AF-Q8WXH0-F29-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F3-model_v6.pdb.gz as AF-Q8WXH0-F3-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F4-model_v6.pdb.gz as AF-Q8WXH0-F4-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F5-model_v6.pdb.gz as AF-Q8WXH0-F5-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F6-model_v6.pdb.gz as AF-Q8WXH0-F6-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F7-model_v6.pdb.gz as AF-Q8WXH0-F7-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F8-model_v6.pdb.gz as AF-Q8WXH0-F8-model_v6.pdb.gz.\n",
            "Extracted and saved AF-Q8WXH0-F9-model_v6.pdb.gz as AF-Q8WXH0-F9-model_v6.pdb.gz.\n",
            "Created zip archive: extracted_files.zip\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_b7fc898a-ef3b-486c-a9ca-d0b173d4410f\", \"extracted_files.zip\", 5981346)"
            ]
          },
          "metadata": {}
        }
      ]
    }
  ]
}